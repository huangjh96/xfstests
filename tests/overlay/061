#! /bin/bash
# FSQA Test No. 061
#
# Test ro/rw fd data inconsistecies
#
# This is a variant of test overlay/016 with mread instead of pread.
# This test requires and enables copy_up_shared feature.
#
# This simple test demonstrates a known issue with overlayfs:
# - process A opens file F for read + mmap
# - process B writes new data to file F
# - process A reads old data from mapped memory
#
#-----------------------------------------------------------------------
#
# Copyright (C) 2016 CTERA Networks. All Rights Reserved.
# Author: Amir Goldstein <amir73il@gmail.com>
#
# This program is free software; you can redistribute it and/or
# modify it under the terms of the GNU General Public License as
# published by the Free Software Foundation.
#
# This program is distributed in the hope that it would be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write the Free Software Foundation,
# Inc.,  51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
#-----------------------------------------------------------------------
#

seq=`basename $0`
seqres=$RESULT_DIR/$seq
echo "QA output created by $seq"

tmp=/tmp/$$
status=1	# failure is the default!
trap "_cleanup; exit \$status" 0 1 2 3 15

_cleanup()
{
	cd /
	rm -f $tmp.*
}

# get standard environment, filters and checks
. ./common/rc
. ./common/filter

# real QA test starts here
_supported_fs overlay
_supported_os Linux
_require_scratch
_require_scratch_feature copy_up_shared
_require_xfs_io_command "open"

rm -f $seqres.full

_scratch_mkfs >>$seqres.full 2>&1

# Create our test files.
lowerdir=$OVL_BASE_SCRATCH_MNT/$OVL_LOWER
mkdir -p $lowerdir
echo "This is old news" > $lowerdir/foo
echo "This is old news" > $lowerdir/bar

_scratch_mount -o copy_up_shared=on


#
# mmap MAP_SHARED|PROT_READ of rofd
# write to rwfd
# read from mapped memory
#
$XFS_IO_PROG -r $SCRATCH_MNT/bar \
	-C "mmap -r 0 16" \
	-C "open $SCRATCH_MNT/bar" \
	-C "pwrite -S 0x61 0 16" \
	-C "mread -v 0 16" \
| _filter_xfs_io | _filter_scratch

status=0
exit
