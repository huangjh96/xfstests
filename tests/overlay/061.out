QA output created by 061
wrote 16/16 bytes at offset 0
XXX Bytes, X ops; XX:XX:XX.X (XXX YYY/sec and XXX ops/sec)
00000000:  61 61 61 61 61 61 61 61 61 61 61 61 61 61 61 61  aaaaaaaaaaaaaaaa
